### PR TITLE
Streamanity extractor

### PR DESCRIPTION
**Pull request by alerikaisattera**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

Added Streamanity support

### CODE REVIEW

File: yt_dlp/extractor/streamanity.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            'uploader': 'Tom Bombadil (Freddy78)',
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+#       Download JSON that contains video information
+        video_info = self._download_json('https://app.streamanity.com/api/video/{}'.format(video_id), video_id)
+
+#       All relevant & useful information
```
Comment (Ashish0804) [inline yt_dlp/extractor/streamanity.py:26]: Use try_get for these, especially for non essentials like description thumbnail and title and uploader

Comment (alerikaisattera) [reply yt_dlp/extractor/streamanity.py:26]: How exactly to use try_get? Isn't title actually essential?

Comment (Ashish0804) [reply yt_dlp/extractor/streamanity.py:26]: You have to import try_get from utils and then do 
try_get(video_info, lambda x: x['data']['video']['author_name'])

Actually I see that ['data']['video'] is used in all fields so just do 

video_info = self._download_json('https://app.streamanity.com/api/video/{}'.format(video_id), video_id)['data']['video'] as the extraction won't happen if this doesn't exist anyways. And then you can use .get() in fields (no need to import tryget in this case)

Comment (Ashish0804) [reply yt_dlp/extractor/streamanity.py:26]: Title isn't essential as it can be none. Only keys required to get the video are essential, rest can be none.

Comment (alerikaisattera) [reply yt_dlp/extractor/streamanity.py:26]: Do I have to use get() in all fields or only non-essential ones? How to use get()? I can't find it in the code

Comment (Ashish0804) [reply yt_dlp/extractor/streamanity.py:26]: You can use get for non essentials. Get is an built-in function for python dicts and you don't have to import it. 
Basically get takes in a key and if it Exists in the dict, then it returns it's value else None or expected value. 

So of a dict is a= { '1':'one'}
Then a.get('1') will return 'one'
a.get('2') will return None
a.get('2', 'two') will return 'two'
And a['2'] will throw error
For more info see 
https://www.w3schools.com/python/ref_dictionary_get.asp

File: yt_dlp/extractor/streamanity.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+class StreamanityIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?streamanity\.com/video/(?P<id>[A-Za-z0-9]+)'
+    _TEST = {
+        'url': 'https://streamanity.com/video/9DFPTnuYi8f2',
+        'info_dict': {
+            'id': '9DFPTnuYi8f2',
+            'ext': 'mp4',
+            'title': 'Bitcoin vs The Lighting Network',
+            'thumbnail': 'https://res.cloudinary.com/streamanity-next/image/upload/v1631475198/thumb/523908e0-e85b-4555-9050-78a8378835f1.png',
+            'description': '',
```
Comment (Ashish0804) [inline yt_dlp/extractor/streamanity.py:16]: Also can you add a test where description is not empty? Of it's big then you can write 
'md5:md5hashofstring' instead of the whole string

File: yt_dlp/extractor/streamanity.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+class StreamanityIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?streamanity\.com/video/(?P<id>[A-Za-z0-9]+)'
+    _TEST = {
+        'url': 'https://streamanity.com/video/9DFPTnuYi8f2',
+        'info_dict': {
+            'id': '9DFPTnuYi8f2',
+            'ext': 'mp4',
+            'title': 'Bitcoin vs The Lighting Network',
+            'thumbnail': 'https://res.cloudinary.com/streamanity-next/image/upload/v1631475198/thumb/523908e0-e85b-4555-9050-78a8378835f1.png',
```
Comment (pukkandan) [inline yt_dlp/extractor/streamanity.py:15]:
```suggestion
            'thumbnail': 're:https://res\.cloudinary\.com/.+\.png',
```

File: yt_dlp/extractor/streamanity.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+        return {
+            'id': video_id,
+            'title': title,
+            'description': description,
+            'uploader': uploader,
+            'is_live': False,
+            'thumbnail': thumbnail,
+            'formats': formats,
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/streamanity.py:49]:
```suggestion
        video_info = self._download_json(
            f'https://app.streamanity.com/api/video/{video_id}', video_id)['data']['video']

        formats = self._extract_m3u8_formats(
            f'https://stream.mux.com/{video_id}.m3u8?token={video_info["token"]}',
            video_id, ext='mp4',m3u8_id='hls', live=False)
        self._sort_formats(formats)

        return {
            'id': video_id,
            'title': video_info['title'],
            'description': video_info.get('description'),
            'uploader': video_info.get('author_name'),
            'is_live': False,
            'thumbnail': video_info.get('thumbnail'),
            'formats': formats,
        }
```

File: yt_dlp/extractor/streamanity.py (hunk 1 above)
Comment (pukkandan) [reply yt_dlp/extractor/streamanity.py]: I've fixed this

File: yt_dlp/extractor/streamanity.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        video_info = self._download_json(
+            f'https://app.streamanity.com/api/video/{video_id}', video_id)['data']['video']
+
+        formats = self._extract_m3u8_formats(
+            f'https://stream.mux.com/{video_id}.m3u8?token={video_info["token"]}',
+            video_id, ext='mp4',m3u8_id='hls', live=False)
```
Comment (pukkandan) [inline yt_dlp/extractor/streamanity.py:28]:
```suggestion
            video_id, ext='mp4', m3u8_id='hls')
```

File: yt_dlp/extractor/streamanity.py (hunk 3 above)
Comment (alerikaisattera) [reply yt_dlp/extractor/streamanity.py]: This version is flawed. Master m3u8 playlist is not "https://stream.mux.com/{video_id}.m3u8?token={video_info["token"]}", but https://stream.mux.com/{play_id}.m3u8?token={video_info["token"]}, where play_id is obtained from JSON at https://app.streamanity.com/api/video/{video_id}

File: yt_dlp/extractor/streamanity.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        }
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        video_info = self._download_json(
+            f'https://app.streamanity.com/api/video/{video_id}', video_id)['data']['video']
+
+        formats = self._extract_m3u8_formats(
+            f'https://stream.mux.com/{video_id}.m3u8?token={video_info["token"]}',
```
Comment (pukkandan) [inline yt_dlp/extractor/streamanity.py]:
```suggestion
            'https://stream.mux.com/%s.m3u8?token=%s' % (video_info['play_id'], video_info['token']),
```

that was my mistake. this should fix it

Comment (alerikaisattera) [reply yt_dlp/extractor/streamanity.py]: I already did

**Comment by alerikaisattera** [comment]
Last thing to do should be adding md5 hashes to the tests. Unfortunately I have to go to sleep, So either I will do it tomorrow, or someone else should do it.

**Comment by alerikaisattera** [comment]
This one is now complete
